openupgrade_scripts/scripts/loyalty/16.0.1.0/pre-migration.py

```diff
@@ -0,0 +1,815 @@
+    ("coupon.reward", "coupon_reward", "discount_apply_on", "discount_applicability"),
+    ("coupon.reward", "coupon_reward", "discount_type", "discount_mode"),
+    (
+        "coupon.reward",
+        "coupon_reward",
+        "discount_specific_product_ids",
+        "discount_product_ids",
+    ),
+    ("coupon.reward", "coupon_reward", "reward_product_quantity", "reward_product_qty"),
+    ("coupon.rule", "coupon_rule", "rule_minimum_amount", "minimum_amount"),
+    ("coupon.rule", "coupon_rule", "rule_min_quantity", "minimum_qty"),
+    ("coupon.rule", "coupon_rule", "rule_products_domain", "product_domain"),
+    (
+        "coupon.rule",
+        "coupon_rule",
+        "minimum_amount_tax_inclusion",
+        "minimum_amount_tax_mode",
+    ),
+]
+_models_renames = [
+    ("coupon.program", "loyalty.program"),
+    ("coupon.reward", "loyalty.reward"),
+    ("coupon.coupon", "loyalty.card"),
+    ("coupon.rule", "loyalty.rule"),
+]
+_tables_renames = [
+    ("coupon_program", "loyalty_program"),
+    ("coupon_reward", "loyalty_reward"),
+    ("coupon_coupon", "loyalty_card"),
+    ("coupon_rule", "loyalty_rule"),
+]
+_xmlids_renames = [
+    (
+        "loyalty.coupon_action",
+        "loyalty.loyalty_card_action",
+    ),
+    (
+        "loyalty.coupon_generate",
+        "loyalty.loyalty_generate_wizard_action",
+    ),
+    (
+        "loyalty.report_coupon_code",
+        "loyalty.report_loyalty_card",
+    ),
+    (
+        "loyalty.coupon_view_form",
+        "loyalty.loyalty_card_view_form",
+    ),
+    (
+        "loyalty.coupon_view_tree",
+        "loyalty.loyalty_card_view_tree",
+    ),
+    (
+        "loyalty.coupon_generate_view_form",
+        "loyalty.loyalty_generate_wizard_view_form",
+    ),
+    (
+        "loyalty.coupon_program_view_form_common",
+        "loyalty.loyalty_program_view_form",
+    ),
+    (
+        "loyalty.coupon_program_view_search",
+        "loyalty.loyalty_program_view_search",
+    ),
+    (
+        "loyalty.coupon_program_view_tree",
+        "loyalty.loyalty_program_view_tree",
+    ),
+    (
+        "sale_loyalty.mail_template_sale_coupon",
+        "loyalty.mail_template_loyalty_card",
+    ),
+    (
+        "gift_card.gift_card_product_50",
+        "loyalty.gift_card_product_50",
+    ),
+]
+
+
+def update_loyalty_program_data(env):
+    # Fill date_to fill: This data takes the values of the following fields:
+    # - validity_duration (coupon.program): in case of coupon programs that generate
+    # coupons with a validity time, it is calculated by taking the creation date of
+    # the program to which the duration defined in the field is added to obtain the
+    # date.
+    # - rule_date_to (coupon.rule): in the case of promotional programs, the date of
+    # rule_date_to is taken.
+    openupgrade.logged_query(
+        env.cr,
+        """
+        ALTER TABLE loyalty_program
+        ADD COLUMN IF NOT EXISTS date_to DATE
```

Comment:
Why wouldn't it exist?

Reply:
Because in v15 this field did not exist in coupon.program(loyalty.program), the validity was calculated with the validity_duration field. When making scripts in pre it is better to make sure if it exists or not before creating it, isn't it?

Reply:
I'd say you're mixing concepts here. In one hand we already had `rule_date_to`, which is the one to dump into this new column (it's true that it was in another relation).

On the other, we've got the **coupon** experation, which is the one that `validity_duration` was setting. But that mechanism doesn't exist anymore.

I'd say that:

- there's nothing to do with `validity_duration`
- you should move `rule_date_to` values to `date_to` in `loyalty_program`

Reply:
But then, in the case of a coupon promotion, how do we establish the validity of the coupon?
For example:
In v15 we have program_type="coupon_program" which is the program that generates coupons and in this we set the validity_duration field.
![image](https://github.com/OCA/OpenUpgrade/assets/118818446/e366a614-2e76-4cc0-8919-a1c4fea72183)
In this view the field rule_date_to cannot be set because in v15 there are two different views for the two types of promotions and in the case of "coupon_program" this field is not in the view.

In v16 we have the same view for all program_types, in this case the counterpart would be program_type="coupons" and we need that validity but in date. 
![image](https://github.com/OCA/OpenUpgrade/assets/118818446/3a1f57e7-ed6c-4485-9071-08b01736fe05)
In the case of coupon programs that come with a past validity date, when migrating them, they would be valid again if a validity date is not calculated from the duration.

Reply:
Existing coupons already have their `expiration_date` set. That's shouldn't change. And as far as I can see, those dates aren't triggered by the dates on the program.

Reply:
It is true that the coupons have the expiry date set in these cases, what had led me to error is that in v16 it is not set when generating the coupons but in the data coming from v15 it will be. Thanks!